### ISSUE TITLE
Add support for switch/match element

### ISSUE THREAD

**Issue by ogoffart** [slint-ui/slint#1307]
Not to be confused with #241which is about expression in bindings.

It is often desirable to choose a different component base on some condition or value

```
if mode == 0 : Foo { ... }
if mode == 1 : Bar  { ... }
```

It would be nice to do 
```
switch mode {
    case 0: Foo { }
    case 1: Bar { }
}
```

Maybe not switch but match, and could be combined with the for model.

```
for data in some_model : match data.type {
    0 : Foo { ... }
    1 : Bar { ... }
    else: Yoo { ... }  // note that this would be ambiguous if else is not a keyword     
}
```

We could also imagine to have an union type with data (similar to rust enums)  
(or have enums with data?)

```
union Shape := {
     rectangle := { h: length, w: length }
     circle := { r: length }
} 

property <[Shape]> model: [ Shape.rectangle{ h: 23px, w: 43px; }, Shape.circle{ r: 4px } ];
for shape in model: match shape {
      Shape.rectangle(rect) => Rectangle { width: rect.w; height: rect.h;  }
      Shape.circle(c) => Rectangle { width: c.r; height: c.r; border-radius: c.r/2; }
}
```

The syntax need to be somehow defined.

**Comment by hunger** [comment]
How involved is it to build powerful (rust-like) match semantics into C++?

**Comment by ogoffart** [comment]
> How involved is it to build powerful (rust-like) match semantics into C++?

If the question is how to lower the proposed `union` into C++ for the C++ api, i guess we can lower that into a C++ variant of different type:

```
union Shape := {
     rectangle := { h: length, w: length }
     circle := { r: length }
} 
```

would become something like

```c++
struct Shape_rectangle { float h; float w; };
struct Shape_circle { float r; };
using Shape = std::variant<Shape_rectangle, Shape_circle>;
```

I guess we have no way around creating types and finding name for them, and using `_` as a separator seems logical.  Perhaps another idea would be to group them in a namespace or something.

**Comment by Tmpod** [comment]
There doesn't seem to be any difference between this issue and #241, is there?

**Comment by ogoffart** [comment]
#241 is for expressions (within property binding or callback handlers), while this issue is about instantiating elements

**Comment by lambdachad** [comment]
This would be excellent to have. Currently I'm having a hard time finding a good way to have per screen state besides having everything local. It would be nice to have `in-out property <Screen> current-screen;` where `Screen` is a tagged union / Rust enum with data support.

What's the current best way to switch screens and assign state to them? Does everything have to be global? I currently have one main app that switches screen based on simple enum.

**Comment by alextechcc** [comment]
Yeah, I agree. The lack of recursive types, variant types / enums makes tree navigation screen structures very difficult without centralizing everything and having a bunch of unset parameters in structs.

**Comment by nichturner** [comment]
I would be interested in attempting to add this

**Comment by 0x6e** [comment]
For context @nichturner is an intern at KDAB, and our initial goal is to a prepare a PR that introduces experimental support for simple match statements, e.g.:

```slint
in property <int> mode: 0;
match self.mode {
    0 : Foo { ... }
    1 : Bar { ... }
    else: Yoo { ... }  // note that this would be ambiguous if else is not a keyword     
}
```

**Comment by 0x6e** [comment]
We made some [design decisions](https://github.com/slint-ui/slint/pull/12021#issuecomment-4721426582) a result of a Mattermost discussion regarding the initial PR #12021 .

**Comment by ogoffart** [comment]
https://github.com/slint-ui/slint/pull/12021  is now merged and i updated the description with a checklist of things before we can mark this as finished.

**Comment by bryce-happel-walton** [comment]
When there is exhaustive checking, it would be nice to be able to do both of the following:

```slint
// equivalent to x := Rect {@children}
x := match y {
 *: Rect {@children}
}
```
 Asking for `@children` may be too much, but if the match is exhaustive, then the component is guaranteed. Whatever infrastructure would allow this would also allow referencing guaranteed properties of parents too, such as position and size.
